regulator: X-Power AXP192 PMIC support

imliubo opened the pull request:
This update simply supports X-Powers AXP192 PMIC, tested with M5Stack StickC_Plus development board, supports setting BUCK1,2,3 and LDO2,3 startup voltage.

kartben commented on boards/xtensa/m5stickc_plus/m5stickc_plus_defconfig:
```diff
@@ -20,3 +20,5 @@ CONFIG_GEN_IRQ_VECTOR_TABLE=n
 
 CONFIG_I2C=y
 CONFIG_CLOCK_CONTROL=y
+CONFIG_REGULATOR=y
+CONFIG_REGULATOR_AXP192=y
```
not sure this should be here as you haven't declared the regulator in the main devicetree but only in the overlay? probably be best to keep it here but also add a "default" definition for the regulator, with everything turned off/set to sensible default settings, and then show how to override with settings appropriate for turning on backlight etc. in the LVGL sample overlay?

gmarull replied:
redundant

kartben commented:
@imliubo this is very cool, thank you for the work on this already! Also thanks for using the opportunity of this PR to clean-up the m5stickc_plus device tree.
It would be great if you could make all the changes be separate commits, since some of them are not strictly speaking related to AXP192. You will need to rebase/force-push when that's done. I.e have the following commits
- clean-up of the M5stick devicetree
- add driver
- enable LVGL sample for m5stickCplus

kartben commented on dts/bindings/regulator/xpowers,axp192-regulator.yaml:
```diff
@@ -0,0 +1,55 @@
+
+  The PMIC has 3 buck converters and 3 LDOs.
+  The regulators need to be defined as child nodes, strictly following the
+  DC1,2,3 LDO1,2,3 node names. For
```
remove unecessary line break

kartben commented on dts/bindings/regulator/xpowers,axp192-regulator.yaml:
```diff
@@ -0,0 +1,55 @@
+# SPDX -License-Identifier: Apache-2.0
+
+description: |
+  X-Powers AXp192 PMIC
```
```suggestion
  X-Powers AXP192 PMIC
```

imliubo commented:
@kartben Thanks for review, first time work with this awesome project, I will clean my commit and push again, if have any problem please let me know.

imliubo committed:
regulator: X-Powers AXP192 PMIC support

X-Powers AXP192 PMIC support
1. enable/disable output support
2. count/list voltage support
3. set/get voltage support
All functions support BUCK1,2,3, LDO2,3 and LDOIO0.

Signed-off-by: imliubo <imliubo@makingfun.xyz>

imliubo commented:
@kartben Commit clean-up and force-push done, pls review, thanks for your suggestion, it help me a lot :)

IcingTomato commented:
I will be assisting iamliubo with the code fixing and interpretation.

gmarull reviewed (changes requested):
please check my comments, driver seems to have some design issues.

aasinclair commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,230 @@
+		return 0;
+	case AXP192_SOURCE_LDO2:
+		reg_value |= (((init_uv - 180000) / 100000) & 0x0F) << 4;
+		i2c_reg_write_byte_dt(&pconfig->i2c, AXP192_REG_LDO23_OUT_VOL, reg_value);
```
You may be able to use i2c_reg_update_byte_dt here.
You should also consider using the linear_range API for calculating the correct register value.

aasinclair commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,230 @@
+	return ret;
+}
+
+static const struct regulator_driver_api api = {.set_voltage = regulator_axp192_set_voltage};
```
I think you should support more of the interface than just set_voltage.
If you use linear_range to define the available voltages, getvoltage, count_voltages and list_voltage will be fairly easy (see the other regulator drivers for inspiration).
Can the regulators be enabled/disabled by SW?

imliubo replied:
> I think you should support more of the interface than just set_voltage.
If you use linear_range to define the available voltages, getvoltage, count_voltages and list_voltage will be fairly easy (see the other regulator drivers for inspiration).

Ok, i will support more interface. 

> Can the regulators be enabled/disabled by SW?

Yes, BUCK1,2,3 and LDO2,3 can be en/dis by softeware.

imliubo commented on include/zephyr/dt-bindings/regulator/axp192.h:
```diff
@@ -0,0 +1,16 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_REGULATOR_AXP192_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_REGULATOR_AXP192_H_
+
+/**
+ * @defgroup regulator_axp192 AXP192 Devicetree helpers.
+ * @ingroup regulator_interface
+ * @{
+ */
+
+#endif /* ZEPHYR_INCLUDE_DT_BINDINGS_REGULATOR_AXP192_H_*/
```
ok

imliubo commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,230 @@
+
+int regulator_axp192_set_voltage(const struct device *dev, int32_t min_uv, int32_t max_uv)
+{
+	/* MIN and MAX voltage are not used */
+	ARG_UNUSED(min_uv);
+	ARG_UNUSED(max_uv);
+
```
I didn't quite understand the usage of this api before, but I understand after seeing this [link](https://docs.zephyrproject.org/latest/samples/shields/npm1300_ek/doc/index.html#regulators), I will deal with this.

imliubo commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,230 @@
+	uint8_t config_bit[3] = {7, 2, 3};
+	uint8_t reg_value = 0;
+
+	i2c_reg_read_byte_dt(&pconfig->i2c, AXP192_REG_DC123_LDO23_CTL, &reg_value);
```
Wow, this API is very convenient, sorry I didn't know this before :)

imliubo commented:
@aasinclair @gmarull Thanks for review. I got flu just feel better now, today I will be dealing with the above great advice, thank you so much!

imliubo commented:
Hi @aasinclair @gmarull @kartben , the driver code refactoring is complete, request to review again.
Now support those API:
1. enable/disable output
2. count/list voltage
3. set/get voltage

Here is sample test log with shell command:
```bash
*** Booting Zephyr OS build zephyr-v3.3.0-4004-g52eaa0694606 ***
uart:~$ regulator disable LDO2  # m5stackc_plus screen backlight turn off
uart:~$ regulator enable LDO2  # m5stackc_plus screen backlight turn on
uart:~$ regulator vget LDO2
3.300 V
uart:~$ regulator vset LDO2 2.95V 3.1V
uart:~$ regulator vget LDO2
3.000 V
uart:~$ regulator vget BUCK2
2.275 V
uart:~$ regulator vget BUCK3
3.300 V
uart:~$ regulator vget BUCK1
3.300 V
uart:~$ regulator vget LDO3
3.300 V
uart:~$ regulator vlist LDO2
1.800 V
1.900 V
2.000 V
2.100 V
2.200 V
2.300 V
2.400 V
...
uart:~$ 
```

But I found a problem,the first communication between AXP192 and esp32 always fails, here is some detail.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/26544331/24430658-3547-42af-9ad2-0f34f7523d4a)

[code link](https://github.com/zephyrproject-rtos/zephyr/pull/57959/files#diff-83efb9bcf04f877774f4c116d20a5c408f19c45ae2c2a40c774553bd630a9a75R344)

When initializing, the first time read the registers of AXP192, it always fails(any registers always fails). I am not sure whether this is the problem of the I2C driver or the problem of the AXP192 driver. If have any suggestions, please let me know.

aasinclair commented:
> When initializing, the first time read the registers of AXP192, it always fails(any registers always fails). I am not sure whether this is the problem of the I2C driver or the problem of the AXP192 driver. If have any suggestions, please let me know.

Have you got a scope to look at this more closely?
If so take a close look at SCL before the first transaction.

I have seen this when the I2C pull-ups are not enabled at initialisation:
https://github.com/zephyrproject-rtos/zephyr/pull/58081

gmarull commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,408 @@
+static const struct linear_range buck2_range = LINEAR_RANGE_INIT(700000, 25000, 0U, 63U);
+static const struct linear_range ldo23_range = LINEAR_RANGE_INIT(1800000, 100000, 0U, 15U);
+
+static int buck_output_enable(const struct device *dev, uint8_t chan, uint8_t enable)
```
`bool enable`

gmarull commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,408 @@
+	const struct regulator_axp192_pconfig *pconfig = dev->config;
+	uint8_t reg_value;
+
+	i2c_reg_read_byte_dt(&pconfig->i2c, AXP192_REG_POWER_STATUS, &reg_value);
+	i2c_reg_read_byte_dt(&pconfig->i2c, AXP192_REG_CHIP_ID, &reg_value);
```
error handling missing here

gmarull commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,408 @@
+		return 0;
+	}
+
+	return -ENODEV;
```
maybe use `-EINVAL` here

gmarull commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,408 @@
+		return -ENODEV;
+	}
+
+	ret = regulator_common_init(dev, false);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return ret;
```
`return regulator_common_init(...`

gmarull commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,408 @@
+
+int regulator_axp192_init(const struct device *dev)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	int ret = 0;
+
+	if (!device_is_ready(config->p)) {
```
missing here: `regulator_common_data_init(dev);`

gmarull commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,408 @@
+#define AXP192_REG_BAT_DCHG_CUR      0x7C
+
+/* X-Powers AXP192 voltage sources */
+enum axp192_voltage_sources {
+	AXP192_SOURCE_BUCK1,
+	AXP192_SOURCE_BUCK2,
+	AXP192_SOURCE_BUCK3,
+	AXP192_SOURCE_LDO1,
+	AXP192_SOURCE_LDO2,
+	AXP192_SOURCE_LDO3
```
note: datasheet seems to advertise 4 LDOs:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/25011557/090c11f4-cba8-4242-829a-e654f1b3c198)

imliubo replied:
Added

gmarull commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,408 @@
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_pconfig *pconfig = config->p->config;
+	uint8_t mask_bit[6] = {0, 4, 1, 7, 2, 3};
```
note: can you use defines here? this looks like magic numbers

gmarull commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,408 @@
+	}
+}
+
+static int buck_set_voltage(const struct device *dev, uint8_t chan, int32_t min_uv, int32_t max_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_pconfig *pconfig = config->p->config;
+	uint16_t idx;
+	int ret;
+
+	if (chan == AXP192_SOURCE_BUCK2) {
+		ret = linear_range_get_win_index(&buck2_range, min_uv, max_uv, &idx);
+	} else {
+		ret = linear_range_get_win_index(&buck13_range, min_uv, max_uv, &idx);
+	}
+
+	if (ret == -EINVAL) {
+		return ret;
+	}
+
+	switch (chan) {
+	case AXP192_SOURCE_BUCK1:
+		return i2c_reg_write_byte_dt(&pconfig->i2c, AXP192_REG_BUCK1_OUT_VOL, idx);
+	case AXP192_SOURCE_BUCK2:
+		return i2c_reg_write_byte_dt(&pconfig->i2c, AXP192_REG_BUCK2_OUT_VOL, idx);
+	case AXP192_SOURCE_BUCK3:
+		return i2c_reg_write_byte_dt(&pconfig->i2c, AXP192_REG_BUCK3_OUT_VOL, idx);
+	default:
+		return -ENODEV;
+	}
+}
```
suggestion here: this function could take a register and an index, determined by the caller function (which already knows if it's BUCK1/2/3. This would avoid checking source/chan 3 times.

aasinclair commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,408 @@
+	}
+}
+
+static int initial_set_voltage(const struct device *dev, int32_t init_uv)
```
This function is not used.

gmarull commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,408 @@
+	uint8_t idx;
+	int ret;
+
+	switch (chan) {
+	case AXP192_SOURCE_BUCK1:
+		ret = i2c_reg_read_byte_dt(&pconfig->i2c, AXP192_REG_BUCK1_OUT_VOL, &idx);
+		break;
+	case AXP192_SOURCE_BUCK2:
+		ret = i2c_reg_read_byte_dt(&pconfig->i2c, AXP192_REG_BUCK2_OUT_VOL, &idx);
+		break;
+	case AXP192_SOURCE_BUCK3:
+		ret = i2c_reg_read_byte_dt(&pconfig->i2c, AXP192_REG_BUCK3_OUT_VOL, &idx);
+		break;
+	default:
+		return -ENODEV;
+	}
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (chan == AXP192_SOURCE_BUCK2) {
+		return linear_range_get_value(&buck2_range, idx, volt_uv);
+	} else {
+		return linear_range_get_value(&buck13_range, idx, volt_uv);
+	}
```
same here

gmarull commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,408 @@
+	}
+
+	if (chan == AXP192_SOURCE_LDO2) {
+		return linear_range_get_value(&ldo23_range, (idx & 0xf0) >> 4, volt_uv);
```
magic mask/shift

gmarull commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,408 @@
+	case AXP192_SOURCE_BUCK2:
+	case AXP192_SOURCE_BUCK3:
+		return buck_output_enable(dev, config->source, true);
+	case AXP192_SOURCE_LDO1:
+		return -ENOTSUP;
```
this is an always-on LDO, so function needs to return 0.

gmarull commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,408 @@
+		return -ENODEV;
+	}
+
+	ret = regulator_common_init(dev, false);
```
LDO1 is enabled, so false is wrong for that case.

aasinclair reviewed (changes requested):
That's starting to look a lot better :-).
Remember you need to sign your commits:
https://docs.zephyrproject.org/latest/contribute/guidelines.html

imliubo committed:
board: clean-up of the M5StickC-Plus devicetree

Tweak SPI3 and I2C0 default pin.

Signed-off-by: imliubo <imliubo@makingfun.xyz>

imliubo committed:
samples: enable LVGL sample for m5stickc-plus board

M5Stickc-plus board has LCD screen, compatible with this LVGL sample.

Signed-off-by: imliubo <imliubo@makingfun.xyz>

imliubo committed:
board: declare axp192 pmic in m5stickc-plus board devicetree

Declare axp192 pmic and set default voltage value for m5stickc-plus board,
the default voltage value depend on which device connect to axp192 pmic.

Signed-off-by: imliubo <imliubo@makingfun.xyz>

imliubo commented:
> > When initializing, the first time read the registers of AXP192, it always fails(any registers always fails). I am not sure whether this is the problem of the I2C driver or the problem of the AXP192 driver. If have any suggestions, please let me know.
> 
> Have you got a scope to look at this more closely? If so take a close look at SCL before the first transaction.
> 
> I have seen this when the I2C pull-ups are not enabled at initialisation: #58081

Works!

kartben commented on boards/xtensa/m5stickc_plus/m5stickc_plus.dts:
```diff
@@ -80,16 +80,65 @@
+				regulator-init-microvolt = <3300000>;
+				regulator-boot-on;
+			};
+
+			pmic_axp192_buck2: BUCK2 {
+				/* 700mV~2275mV@25mV/step, max output 1600mA */
+				regulator-init-microvolt = <2275000>;
+				regulator-boot-on;
+			};
+
+			pmic_axp192_buck3: BUCK3 {
+				/* 700mV~3500mV@25mV/step, max output 700mA */
+				regulator-init-microvolt = <3300000>;
+				regulator-boot-on;
+			};
+
+			pmic_axp192_ldo1: LDO1 {
+				/* Always on */
+				regulator-always-on;
+			};
+
+			pmic_axp192_ldo2: LDO2 {
+				/* 1800mV~3300mV@100mV/step, max output 200mA */
+				regulator-init-microvolt = <3300000>;
+				regulator-boot-on;
+			};
+
+			pmic_axp192_ldo3: LDO3 {
+				/* 1800mV~3300mV@100mV/step, max output 200mA */
+				regulator-init-microvolt = <3300000>;
+				regulator-boot-on;
+			};
+
+			pmic_axp192_ldoio0: LDOIO0 {
+				/* 1800mV~3300mV@100mV/step, max output 50mA */
+				regulator-init-microvolt = <3300000>;
+				regulator-boot-on;
+			};
+		};
+	};
```
Wouldn't it make sense to have here only the settings that are generic (if any? if not, just declare the regulator, with no other setting) and then "turn on" the TFT backlight and TFT controller in the overlay of the LVGL sample?
I think it's important that people understand at first glance that there is an easy way to completely cutoff all LCD-related components by using the regulator. It also helps have the "default" devicetree get closer to demonstrate minimal power consumption (although my understanding is that power-management is not well/fully supported yet for the ESP-32 port of Zephyr).

gmarull replied:
You should actually make use of `xxx-supply = <&regulator>;` in the display, and be the display driver who controls the regulator on/off state.

imliubo replied:
Sorry, not sure how to write this

gmarull commented on samples/subsys/display/lvgl/boards/m5stickc_plus.overlay:
```diff
@@ -0,0 +1,43 @@
+	};
+};
+
+&spi3 {
+	st7789v: st7789v@0 {
+		compatible = "sitronix,st7789v";
+		spi-max-frequency = <20000000>;
+		reg = <0>;
+		cmd-data-gpios = <&gpio0 23 GPIO_ACTIVE_LOW>;
+		reset-gpios = <&gpio0 18 GPIO_ACTIVE_LOW>;
+		vcom = <0x28>;
+		gctrl = <0x35>;
+		vrhs = <0x10>;
+		vdvs = <0x20>;
+
+		/* Landscape */
+		width = <240>;
+		height = <135>;
+		x-offset = <40>;
+		y-offset = <52>;
+		mdac = <0xA8>;
+
+		gamma = <0x01>;
+		colmod = <0x55>;
+		lcm = <0x0C>;
+		porch-param = [0C 0C 00 33 33];
+		cmd2en-param = [5A 69 02 01];
+		pwctrl1-param = [A4 A1];
+		pvgam-param = [D0 00 02 07 0A 28 32 44 42 06 0E 12 14 17];
+		nvgam-param = [D0 00 02 07 0A 28 31 54 47 0E 1C 17 1B 1E];
+		ram-param = [00 E0];
+		rgb-param = [CD 08 14];
+	};
```
As far as I can see this board provides a display, so please, move this to board level dts.

imliubo replied:
ok

gmarull commented on dts/bindings/regulator/xpowers.axp192-regulator.yaml:
```diff
@@ -0,0 +1,55 @@
+        - regulator-min-microvolt
+        - regulator-max-microvolt
+        - regulator-init-microvolt
+        - regulator-min-microamp
+        - regulator-max-microamp
```
not supported

gmarull commented on boards/xtensa/m5stickc_plus/m5stickc_plus.dts:
```diff
@@ -82,6 +82,56 @@
+				regulator-boot-on;
+			};
+
+			pmic_axp192_ldo1: LDO1 {
+				/* Always on */
+				regulator-always-on;
+			};
```
see my comments regarding LDO1. Regulator DT props are meant to configure the behavior, not to specify its props.

gmarull reviewed (changes requested):
Things look pretty good now, see my comments

aasinclair commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,435 @@
+static const struct linear_range buck2_range = LINEAR_RANGE_INIT(700000, 25000, 0U, 63U);
+static const struct linear_range ldo23_io0_range = LINEAR_RANGE_INIT(1800000, 100000, 0U, 15U);
+
+static int buck_output_enable(const struct device *dev, uint8_t reg_addr, uint8_t ctl_bit,
```
buck_output_enable and ldo_output_enable are identical.
You could call i2c_reg_update_byte_dt directly from regulator_axp192_enable

imliubo commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,435 @@
+		return linear_range_values_count(&buck13_range);
+	case AXP192_SOURCE_BUCK2:
+		return linear_range_values_count(&buck2_range);
+	case AXP192_SOURCE_LDO1:
+		return -ENOTSUP;
```
I saw LDO1 has 4 fixed voltage values can be adjusted, the output cannot be turned off, I am not found the voltage control register in datasheet, but we ask our supply, waiting them tell us the right register to control the voltage, if has this register i will add fully support of LDO1.

imliubo commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -0,0 +1,435 @@
+	bool is_enabled = false;
+	int ret = 0;
+
+	regulator_common_data_init(dev);
+
+	if (!device_is_ready(config->p)) {
+		return -ENODEV;
+	}
+
+	regulator_common_data_init(dev);
```
emmm, i didn't remember add this for twice

gmarull commented:
please add a device entry in `tests/drivers/build_all/regulator` as well.

MrMarteng commented on boards/xtensa/m5stickc_plus/m5stickc_plus.dts:
```diff
@@ -80,16 +80,65 @@
 	pinctrl-names = "default";
+
+	pmic_xpowers_axp192: pmic@34 {
+		compatible = "xpowers,axp192";
```
Where is this driver located?

MrMarteng commented on boards/xtensa/m5stickc_plus/m5stickc_plus.dts:
```diff
@@ -80,16 +80,65 @@
+			pmic_axp192_buck1: BUCK1 {
+				/* 700mV~3500mV@25mV/step, max output 1200mA */
+				regulator-init-microvolt = <3300000>;
+				regulator-boot-on;
```
I am not sure if the `regulator-boot-on` option is causing the intended behavior.
```
 regulator-boot-on:
    type: boolean
    description: |
      bootloader/firmware enabled regulator.
      It's expected that this regulator was left on by the bootloader.
      If the bootloader didn't leave it on then OS should turn it on
      at boot but shouldn't prevent it from being turned off later.
      This property is intended to only be used for regulators where
      software cannot read the state of the regulator.
```
Please correct me if I am wrong, but to my knowledge the m5stack boot loader does not feature regulator support, and therefore it may start without being powered on.

imliubo replied:
emmmm, I am not sure the bootloader on zephyr port, but i think zephyr also has some bootloader stage when boot on? 
Like this:
esp32 bootloader(1st stage) -> zephyr bootlader(2nd stage) -> zephyr OS
So this regulator-boot-on option is finish on zephyr bootlader stage?

Just my guess, i am newbie of zephyr.

MrMarteng replied:
Sorry, my comment was maybe misleading. To my understanding the `regulator-boot-on` option tells the common regulator API that it may assume the regulator is already on. This could lead to the behaviour that the according regulator is not being enabled again. As the description says this option is intended to be used for regulators where software cannot read the state of the regulator.
I suppose the main reason this is still working in your case, is that you always pass "false" to the `regulator_common_init()` function, so that the common regulator api will always enable your regulator.

I doubt that any of the existing bootloaders already has support to enable the regulator.

imliubo commented on boards/xtensa/m5stickc_plus/m5stickc_plus.dts:
```diff
@@ -80,16 +80,65 @@
 	pinctrl-names = "default";
+
+	pmic_xpowers_axp192: pmic@34 {
+		compatible = "xpowers,axp192";
```
[Here](https://github.com/zephyrproject-rtos/zephyr/pull/57959/files#diff-83efb9bcf04f877774f4c116d20a5c408f19c45ae2c2a40c774553bd630a9a75R1-R435)

imliubo commented:
I am very sorry for not continuing this PR for a long time, i am a little busy those days, I saw @MrMarteng is also doing the same work (doing PMIC support for M5Stack devices, thanks very much!), which PR should we continue to prevent duplication of work? I see @MrMarteng PR has mfd layer support, that means this 2 RP has some a little different. 

Any ideas? @kartben @aasinclair @gmarull

MrMarteng commented on boards/xtensa/m5stickc_plus/m5stickc_plus.dts:
```diff
@@ -80,16 +80,65 @@
+				regulator-boot-on;
+			};
+
+			pmic_axp192_ldoio0: LDOIO0 {
```
I don't have the schematic of the M5StickC, but are you sure, that this is used as an output?
On M5Stack this pin is used as an GPIO Input (VBUS-Detection)

imliubo replied:
This pin(GPIO0) has different function for different, for StickCPlus it‘s used for MIC power supply.

<img width="734" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/26544331/5bdcff23-9e79-4fc3-941b-43b712efe087">

MrMarteng commented:
> I am very sorry for not continuing this PR for a long time, i am a little busy those days, I saw @MrMarteng is also doing the same work (doing PMIC support for M5Stack devices, thanks very much!), which PR should we continue to prevent duplication of work? I see @MrMarteng PR has mfd layer support, that means this 2 RP has some a little different.
> 
> Any ideas? @kartben @aasinclair @gmarull


Hi @imliubo, @gmarull, @aasinclair, @kartben

first of all, I would like to apologise, that I have created a second PR for axp192 support. If I would have seen this PR before I would have collaborated on it. Nevertheless I have put a lot of effort in the regulator implementation and already started to work on other features of this PMIC, as the M5Stack Core2 module uses even more features of this IC (e.g. LCD reset gpio signal).

@gmarull mentioned that imliubo's PR is more advanced in review stage, but I am not sure if this actually is the case. So I would like to make a case for proceeding with my PR (#58809) regarding the regulator implementation. Here are some reasons that speak for it from my point of view:

*  My variant includes additional functionality:
    * I added a feature to disable regulators on startup (still unsure of this is the right approach, @gmarull your feedback would be very helpful here)
    * Workmode control for DCDC1-3 is implemented
    * Current limits are published via get_current_limit() interface.
    * I added the MFD layer as proposed by @aasinclair for other PMICs already. This seems to be an approach that just recently has been introduced. 
* Also from a formal point of view, I believe that my PR can be completed faster:
    * My PR is dedicated to regulator functionality only
    * it is based on a more recent main version. I already see some slight conflicts with imliubos PR (e.g. the x-power vendor specified has already been added by another developer within the last two weeks,... )
    * it already passes all required CI tests

Moreover, I already started working on the GPIO part for this IC, which also includes extending functionality in my MFD layer. According discussions are ongoing in discord. I plan to create a PR for the GPIO driver asap.

@imliubo , if you are ok with this, I would recommend to proceed with my PR. But please let’s connect on Discord (username: marteng), to align on the further proceeding to get M5Stack Core2 support into Zephyr. I just joined the community a week ago and I mostly available via discord.

kartben commented:
> @imliubo , if you are ok with this, I would recommend to proceed with my PR. But please let’s connect on Discord (username: marteng), to align on the further proceeding to get M5Stack Core2 support into Zephyr. I just joined the community a week ago and I mostly available via discord.

+1, I agree that your PR is probably more complete at this point. Maybe this one could be refocused on M5StickC-Plus specifically, updating its Devicetree to use your new driver when ready, and contributing the LVGL sample overlay?

imliubo commented:
@MrMarteng Agree, it may be better to continue to promote your PR.This PR don't have MFD support and no GPIO function support.Your PR support is better and more powerful.And thank you very much for your support of M5 devices.  When you PR is ready, I may support other M5Stack develop boards with you AXP192 support.

By the way,we use some new PMIC(AXP2101) for our new board, it's not compatible with AXP192,i think i can send you a sample to test and maybe you can help us support this new PMIC in zephyr for our new board(CoreS3)? I will contact you on Monday.

@kartben I think continue with @MrMarteng PR is better for long-term think. And as our(M5Stack), we can support a new M5CoreS3, hope it can help @MrMarteng do AXP2101 support.

MrMarteng commented:
Thank you. I am happy to support you on m5stack device integrating into Zephyr. AXP2101 looks more complex, but once AXP192 support is finished, including fuel gauge and sensors, that should not be a problem.
Looking forward to meet you next week.

gmarull commented:
alternative merged https://github.com/zephyrproject-rtos/zephyr/pull/58809
